### PR TITLE
Collection expressions: use iteration type for convertibility of element types

### PR DESCRIPTION
**Pull request by cston**
Updates following [LDM-2024-01-08](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-01-08.md).

### CODE REVIEW

**Review by AlekseyTs — APPROVED** [review]

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review.

File: proposals/csharp-12.0/collection-expressions.md [hunk 1]
```diff
@@ -102,37 +102,28 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
...
+* A *struct* or *class type* that implements `System.Collections.IEnumerable`
+* An *interface type*:
+  * `System.Collections.Generic.IEnumerable<T>`
+  * `System.Collections.Generic.IReadOnlyCollection<T>`
+  * `System.Collections.Generic.IReadOnlyList<T>`
+  * `System.Collections.Generic.ICollection<T>`
+  * `System.Collections.Generic.IList<T>`
+
+The implicit conversion exists for the type if the following hold:
+* The type has an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `U` where the iteration type is determined from a `GetEnumerator` instance method or enumerable interface, not from extension methods.
```
Comment (AlekseyTs) [inline proposals/csharp-12.0/collection-expressions.md:120]: > is determined from a `GetEnumerator` instance method or enumerable interface, not from extension methods.

I think this was decided exclusively for "CollectionBuilder" types, but it looks like all other types already satisfy this requirement. I am not suggesting any chnages, just want to make sure the more general wording is intentional.

**Comment by AlekseyTs** [comment]
Consider adding a links to Mondays LDM [notes](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-01-08.md) to the "Design meetings" section.

**Comment by AlekseyTs** [comment]
For each resolved design question, consider marking it as resolved, adding information about what was the resolution (approved as is, approved with modifications, etc.), and referencing the relevant section from the LDM notes.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 8)

**Comment by WeihanLi** [comment]
Think we should mention the changes somewhere since this is a breaking change.

My collection expression sample code break when built with a .NET 9 SDK, works with .NET 8 SDK

https://github.com/WeihanLi/SamplesInPractice/commit/81cbd08aad09428099a27cbb4b07ae21474c9375

**Pull request merged**